src/Tools/dotnet-dsrouter/USBMuxTcpClientRouterFactory.cs

```diff
@@ -0,0 +1,492 @@
+using Microsoft.Diagnostics.NETCore.Client;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Diagnostics.Tools.DiagnosticsServerRouter
+{
+    internal class USBMuxInterop
+    {
+        public const string CoreFoundationLibrary = "/System/Library/Frameworks/CoreFoundation.framework/CoreFoundation";
+        public const string MobileDeviceLibrary = "/System/Library/PrivateFrameworks/MobileDevice.framework/MobileDevice";
+        public const string LibC = "libc";
+
+        public const int EINTR = 4;
+
+        public enum AMDeviceNotificationMessage : uint
+        {
+            None = 0,
+            Connected = 1,
+            Disconnected = 2,
+            Unsubscribed = 3
+        }
+
+        public struct AMDeviceNotificationCallbackInfo
+        {
+            public AMDeviceNotificationCallbackInfo(IntPtr device, AMDeviceNotificationMessage message)
+            {
+                this.am_device = device;
+                this.message = message;
+            }
+
+            public IntPtr am_device;
+            public AMDeviceNotificationMessage message;
+        }
+
+        public delegate void DeviceNotificationDelegate(ref AMDeviceNotificationCallbackInfo info);
+
+#region MobileDeviceLibrary
+        [DllImport(MobileDeviceLibrary)]
+        public static extern uint AMDeviceNotificationSubscribe(DeviceNotificationDelegate callback, uint unused0, uint unused1, uint unused2, out IntPtr context);
+
+        [DllImport(MobileDeviceLibrary)]
+        public static extern uint AMDeviceNotificationUnsubscribe(IntPtr context);
+
+        [DllImport(MobileDeviceLibrary)]
+        public static extern uint AMDeviceConnect(IntPtr device);
+
+        [DllImport(MobileDeviceLibrary)]
+        public static extern uint AMDeviceDisconnect(IntPtr device);
+
+        [DllImport(MobileDeviceLibrary)]
+        public static extern uint AMDeviceGetConnectionID(IntPtr device);
+
+        [DllImport(MobileDeviceLibrary)]
+        public static extern int AMDeviceGetInterfaceType(IntPtr device);
+
+        [DllImport(MobileDeviceLibrary)]
+        public static extern uint USBMuxConnectByPort(uint connection, ushort port, out int socketHandle);
+#endregion
+#region CoreFoundationLibrary
+        [DllImport(CoreFoundationLibrary)]
+        public static extern void CFRunLoopRun();
+
+        [DllImport(CoreFoundationLibrary)]
+        public static extern void CFRunLoopStop(IntPtr runLoop);
+
+        [DllImport(CoreFoundationLibrary)]
+        public static extern IntPtr CFRunLoopGetCurrent();
+#endregion
+#region LibC
+        [DllImport(LibC, SetLastError = true)]
+        public static extern unsafe int send(int handle, byte* buffer, IntPtr length, int flags);
+
+        [DllImport(LibC, SetLastError = true)]
+        public static extern unsafe int recv(int handle, byte* buffer, IntPtr length, int flags);
+
+        [DllImport(LibC, SetLastError = true)]
+        public static extern int close(int handle);
+#endregion
```

Comment:
Do we care about wrapping any exceptions that might be thrown by these methods? I don't see try/catch blocks around usage of these. If the DLL fails to be found or other similar errors occur, you may end up with an ugly stack trace that is difficult to diagnose.

Reply:
I decided to do try/catch on first usage of the AM* methods, but didn't put it around each and one of them, reasoning was, if we fail to load, it will happen on first call, and then we won't continue code paths using the library, if others fail that is due to a change in exports, that is a breaking change, and that could/should only happen for the private framework and if that happens there will be other software broken in the tooling (not just this).

Reply:
If we have issues at some point in future with this, this code is only enabled using --forward-port argument and if that is not used, regular TCP/IP client will be used as normal, or possible to run using other dsrouter modes, so there are options available until we could potentially fix any underlying change that breaks this scenario.